File: backend/app/routers/songs.py
from fastapi import APIRouter, Depends

from app.dependencies import require_role
from app.schemas.song import SongCreate, SongItem, SongUpdate
from app.schemas.song_section import SongSectionItem, SongSectionsUpdate
from app.services import song_section_service, song_service

router = APIRouter(prefix="/api/v1/songs", tags=["songs"])


@router.get("", response_model=list[SongItem])
def list_songs(q: str | None = None):
    return song_service.list_songs(q)


@router.post(
    "",
    response_model=SongItem,
    status_code=201,
    dependencies=[Depends(require_role("leader"))],
)
def create_song(payload: SongCreate):
    return song_service.create_song(payload)


@router.delete(
    "/{song_id}",
    status_code=204,
    dependencies=[Depends(require_role("leader"))],
)
def delete_song(song_id: int):
    """곡 마스터 삭제. 어떤 콘티에도 배치돼 있지 않은 곡만 지울 수 있다."""
    song_service.delete_song(song_id)


@router.patch(
    "/{song_id}",
    response_model=SongItem,
    dependencies=[Depends(require_role("leader"))],
)
def update_song(song_id: int, payload: SongUpdate):
    return song_service.update_song(song_id, payload)


@router.get(
    "/{song_id}/sections",
    response_model=list[SongSectionItem],
    dependencies=[Depends(require_role("member"))],
)
def list_song_sections(song_id: int):
    """곡의 구간별 가사. 저작권 있는 콘텐츠라 로그인(member 이상)해야 볼 수 있다."""
    return song_section_service.list_sections(song_id)


@router.put(
    "/{song_id}/sections",
    response_model=list[SongSectionItem],
    dependencies=[Depends(require_role("leader"))],
)
def put_song_sections(song_id: int, payload: SongSectionsUpdate):
    """구간 배열 전체 교체. conti_songs/schedule_assignments와 동일한 전체 교체 패턴."""
    return song_section_service.replace_sections(song_id, payload)

File: frontend/src/api/songs.js
import { apiDelete, apiGet, apiPatch } from "./client";

// 곡 마스터 목록. AI 검수 화면에서 "기존 곡 선택" 드롭다운 옵션으로 쓴다.
export function fetchSongs() {
  return apiGet("/api/v1/songs");
}

// 곡 제목·아티스트 수정. AI 인식 힌트가 곡 마스터를 참고하므로, 잘못 저장된 제목을 여기서 고쳐두면
// 다음 인식부터 그 오타가 재생산되지 않는다.
export function updateSong(songId, payload, password) {
  return apiPatch(`/api/v1/songs/${songId}`, payload, password);
}

// 어떤 콘티에도 배치되지 않은 곡만 지울 수 있다(사용 중이면 서버가 409).
export function deleteSong(songId, password) {
  return apiDelete(`/api/v1/songs/${songId}`, password);
}
